src/tm_devices/drivers/pi/signal_generators/afgs/afg.py

```diff
@@ -0,0 +1,484 @@
+        """
+        del output_signal_path  # Not used in AFGs.
+        self._validate_generated_function(function)
+        # Generate the waveform on the given channel
+        for channel_name in self._validate_channels(channel):
+            source_channel = self.source_channel[channel_name]
+            source_channel.set_function_properties(
+                frequency=frequency,
+                function=function,
+                amplitude=amplitude,
+                offset=offset,
+                burst_count=burst_count,
+                termination=termination,
+                duty_cycle=duty_cycle,
+                polarity=polarity,
+                symmetry=symmetry,
+            )
+            source_channel.set_state(1)
+
+    def generate_burst(self) -> None:
+        """Generate a burst of waveforms by forcing trigger."""
+        self.write("*TRG")
+        self.expect_esr(0)
+
+    def get_waveform_constraints(  # pyright: ignore[reportIncompatibleMethodOverride]
+        self,
+        function: Optional[SignalGeneratorFunctionsAFG] = None,
+        waveform_length: Optional[int] = None,
+        frequency: Optional[float] = None,
+        output_signal_path: Optional[SignalGeneratorOutputPathsBase] = None,
+        load_impedance: LoadImpedanceAFG = LoadImpedanceAFG.HIGHZ,
+    ) -> ExtendedSourceDeviceConstants:
+        """Get the constraints that restrict the waveform to certain parameter ranges.
+
+        Args:
+            function: The function that needs to be generated.
+            waveform_length: The length of the waveform if no function or arbitrary is provided.
+            frequency: The frequency of the waveform that needs to be generated.
+            output_signal_path: The output signal path that was set on the channel.
+            load_impedance: The suggested impedance on the source.
+
+        Returns:
+            A Named Tuple containing a set of parameters and their restricted bounds.
+        """
+        del output_signal_path
+
+        if not function:
+            msg = "AFGs must have a function defined."
+            raise ValueError(msg)
+        (
+            amplitude_range,
+            frequency_range,
+            offset_range,
+            sample_rate_range,
+        ) = self._get_series_specific_constraints(
+            function, waveform_length, frequency, load_impedance
+        )
+
+        return ExtendedSourceDeviceConstants(
+            amplitude_range=amplitude_range,
+            frequency_range=frequency_range,
+            offset_range=offset_range,
+            sample_rate_range=sample_rate_range,
+        )
+
+    ################################################################################################
+    # Private Methods
+    ################################################################################################
+    def _reboot(self) -> None:
+        """Reboot the device."""
+        # TODO: implement
```

Comment:
should these todo's be left in here? or should they raise the NotImplementedError? or they could be removed if the children of this class actually implement it?